dep-graph/client/src/app/shell.tsx

```diff
@@ -0,0 +1,205 @@
+  const environment = useEnvironmentConfig();
+
+  const projectSelected = state.context.selectedProjects.length > 0;
+
+  const { lastPerfReport } = state.context;
+
+  const [selectedProjectId, setSelectedProjectId] = useState<string>(
+    environment.appConfig.defaultProjectGraph
+  );
+
+  function projectChange(projectGraphId: string) {
+    setSelectedProjectId(projectGraphId);
+  }
+
+  useEffect(() => {
+    const { appConfig } = environment;
+
+    const projectInfo = appConfig.projectGraphs.find(
+      (graph) => graph.id === selectedProjectId
+    );
+
+    const fetchProjectGraph = async () => {
+      const project: DepGraphClientResponse =
+        await projectGraphService.getProjectGraph(projectInfo.url);
+
+      const workspaceLayout = project?.layout;
+      send({
+        type: 'initGraph',
+        projects: project.projects,
+        dependencies: project.dependencies,
+        affectedProjects: project.affected,
+        workspaceLayout: workspaceLayout,
+      });
+    };
+    fetchProjectGraph();
+  }, [selectedProjectId, environment]);
+
+  useIntervalWhen(
+    () => {
+      const projectInfo = environment.appConfig.projectGraphs.find(
+        (graph) => graph.id === selectedProjectId
+      );
+
+      const fetchProjectGraph = async () => {
+        const project: DepGraphClientResponse =
+          await projectGraphService.getProjectGraph(projectInfo.url);
+
+        send({
+          type: 'updateGraph',
+          projects: project.projects,
+          dependencies: project.dependencies,
+        });
+      };
+
+      fetchProjectGraph();
+    },
+    5000,
+    environment.watch
+  );
+
+  function downloadImage() {
+    const graph = getGraphService();
+    const data = graph.getImage();
+
+    let downloadLink = document.createElement('a');
+    downloadLink.href = data;
+    downloadLink.download = 'graph.png';
+    // this is necessary as link.click() does not work on the latest firefox
+    downloadLink.dispatchEvent(
+      new MouseEvent('click', {
+        bubbles: true,
+        cancelable: true,
+        view: window,
+      })
+    );
+  }
+
+  return (
+    <>
+      <Sidebar></Sidebar>
+      <div id="main-content" className="flex-grow overflow-hidden">
+        {environment.appConfig.showDebugger ? (
+          <div
+            id="debugger-panel"
+            className="
+          w-auto
+          text-gray-700
+          bg-gray-50
+          border-b border-gray-200
+          p-4
+          flex flex-column
+          items-center
+          justify-items-center
+          gap-4
+        "
+          >
+            <h4 className="text-lg font-bold mr-4">Debugger</h4>
+            <select
+              className="w-auto flex items-center px-4 py-2 border border-gray-300 rounded-md text-sm font-medium text-gray-700 bg-white"
+              data-cy="project-select"
+              onChange={(event) => projectChange(event.target.value)}
```

Comment:
Since this is a top-level component, consider wrapping the callback with `useCallback` and extracting `<select>` into its own component wrapped with `React.memo`.